sklearn/preprocessing/_encoders.py

```diff
@@ -84,6 +118,11 @@ def _fit(self, X, handle_unknown='error'):
             Xi = X_list[i]
             if self.categories == 'auto':
                 cats = _encode(Xi)
+            elif self.categories == 'dtypes':
+                if hasattr(Xi, "cat"):
+                    cats = Xi.cat.categories.values.copy()
+                else:
+                    cats = _encode(Xi)
```

Comment:
An alternative to the if/else check here (the same fore below), is to handle this inside `_encode` (that's the approach that was taken in https://github.com/scikit-learn/scikit-learn/pull/13351). Advantage is to keep some additional complexity out of this already complex code (moving it elsewhere of course ..)